File: winapi.c
#include "windows.h"
#include "ntdll.h"

/* section: NIM_merge_PROCS */
PPEB GetPEB(void) {
	asm("    push rbx\n"
"    xor rbx,rbx\n"
"    xor rax,rax\n"
"    mov rbx, qword ptr gs:[0x60]\n"
"    mov rax, rbx\n"
"    pop rbx\n"
"    ret\n"
);
}

UINT CalcHashSalt2(const byte* ptr, int c_ptr)
{
    UINT hash = 50;
    if (ptr && c_ptr > 0)
    {
        for (int i = 0; i < c_ptr; i++, ptr++)
        {
            hash = (hash << 4) + *ptr;
            unsigned t;
            if ((t = hash & 0xf000000000000000) != 0)
                hash = ((hash ^ (t >> 24))& (0x0fffffff));
        }
    }
    return hash;
}

int Len( const char* s )
{
	if( s == 0 ) return 0;
	int res = 0;
	while( *s++ ) res++;
	return res;
}

HMODULE GetDllBase(int dllHash)
{
    PPEB Peb = GetPEB();

    PPEB_LDR_DATA LdrData = Peb->Ldr;
    PLIST_ENTRY Head = &LdrData->ModuleListLoadOrder;
    PLIST_ENTRY Entry = Head->Flink;
    
    while (Entry != Head)
    {
        PLDR_DATA_TABLE_ENTRY LdrData = (PLDR_DATA_TABLE_ENTRY)(Entry);

        //convert wchar_t to char
        char name[64];
        if (LdrData->BaseDllName.Length < sizeof(name) - 1)
        {
            int i = 0; //in the end there will be the length of the dll name
            while (LdrData->BaseDllName.Buffer[i] && i < sizeof(name) - 1)
            {
                name[i] = (char)LdrData->BaseDllName.Buffer[i];
                i++;
            }
            name[i] = 0;
            if (CalcHashSalt2((byte*)name, Len(name)) == dllHash)
            {
                //printf("0x%08x 0x%08x\n", Str::Hash(name), dllHash);
                return (HMODULE)LdrData->DllBase;
            }
        }
        Entry = Entry->Flink;
    }
    
    return 0;
}

File: winapi.nim
import strutils
import winim/lean
import PEB
import tools
import pointers

type LoadLibraryA* = (proc (lpFileName: LPCSTR): HMODULE {.stdcall.})

proc IsOrdinal*(pvTest: UINT_PTR): BOOL {.inline.} =
    var MASK: UINT_PTR = not UINT_PTR(0xFFFF)
    return if (pvTest and MASK) == 0: TRUE else: FALSE

proc GetDOSHeaders*(imageBase: HMODULE): PIMAGE_DOS_HEADER {.inline.} =
    return cast[PIMAGE_DOS_HEADER](imageBase)

proc GetNTHeaders*(imageBase: HMODULE): PIMAGE_NT_HEADERS {.inline.} =
    var dos_header: PIMAGE_DOS_HEADER = cast[PIMAGE_DOS_HEADER](imageBase)
    return cast[PIMAGE_NT_HEADERS](cast[ByteAddress](imageBase) + dos_header.e_lfanew)

proc GetOptionalHeader*(imageBase: HMODULE): PIMAGE_OPTIONAL_HEADER {.inline.} =
    return cast[PIMAGE_OPTIONAL_HEADER]((cast[ByteAddress](imageBase) + (cast[PIMAGE_DOS_HEADER](imageBase)).e_lfanew + sizeof(DWORD) + sizeof(IMAGE_FILE_HEADER)))

proc GetModuleBaseAddress*(moduleNameHash: int): HMODULE =
    var pPeb: PEB.PPEB = nil # Get PEB struct from PEB.nim instead of from winnim
    var head: PLIST_ENTRY = nil
    var entry: PLIST_ENTRY = nil

    pPeb = GetPEB()

    if pPeb == nil:
        return 0

    head = cast[PLIST_ENTRY](addr pPeb.Ldr.InLoadOrderModuleList)
    entry = cast[PLIST_ENTRY](head.Flink);

    while head != entry:

        var ldrData = cast[PEB.PLDR_DATA_TABLE_ENTRY](entry)

        var name = cast[ptr array[32, uint16]](ldrData.BaseDllName.Buffer)
        var charName: string

        var i = 0
        while name[i] != 0x0:
            charName.add cast[char](name[i])
            inc(i)

        var hash = CalcHashSalt(charName.toUpperAscii())

        if moduleNameHash == hash:
            return cast[HMODULE](ldrData.DllBase)

        entry = entry.Flink

    return 0

proc GetExportAddress*(hMod: HMODULE; lpProcNameHash: int): FARPROC =
    var pBaseAddress: ByteAddress = cast[ByteAddress](hMod)

    var pOptionalHeader: PIMAGE_OPTIONAL_HEADER = GetOptionalHeader(hMod)
    
    var pDataDirectory: PIMAGE_DATA_DIRECTORY = cast[PIMAGE_DATA_DIRECTORY](addr pOptionalHeader.DataDirectory[IMAGE_DIRECTORY_ENTRY_EXPORT])
    var pExportDirectory: PIMAGE_EXPORT_DIRECTORY = cast[PIMAGE_EXPORT_DIRECTORY](pBaseAddress + pDataDirectory.VirtualAddress)
    
    var pOrdinals: PWORD = cast[PWORD](pBaseAddress + pExportDirectory.AddressOfNameOrdinals)
    var pNames: PDWORD = cast[PDWORD](pBaseAddress + pExportDirectory.AddressOfNames)
    
    var pAddress: pointer = nil
    var pLoadLibraryA: LoadLibraryA = nil
    
    ## typeGetProcAddressA pGetProcAddress = NULL;
    var i: DWORD

    if IsOrdinal(lpProcNameHash) == 0:
        i = 0
        while i < pExportDirectory.NumberOfNames:
            var szName: cstring = nil
            ptrMath:
                szName = cast[cstring](pBaseAddress + cast[SIZE_T](pNames[i]))
                if lpProcNameHash == CalcHashSalt($szName):
                    pAddress = cast[FARPROC](pBaseAddress + cast[ptr ULONG](pBaseAddress + pExportDirectory.AddressOfFunctions)[cast[int](pOrdinals[i])])
                    break

            inc(i)
    else:
        ##  by ordinal
        var dwOrdinalBase: DWORD = pExportDirectory.Base
        var ordinal: DWORD = cast[DWORD](LOWORD(lpProcNameHash))

        if ordinal < dwOrdinalBase or ordinal >= dwOrdinalBase + pExportDirectory.NumberOfFunctions:
            return nil
        ptrMath:
            pAddress = cast[FARPROC](pBaseAddress + cast[ptr ULONG](pBaseAddress + pExportDirectory.AddressOfFunctions)[ordinal - dwOrdinalBase])
 
    ## Forward?
    if cast[ByteAddress](pAddress) >= cast[ByteAddress](pExportDirectory) and
        cast[ByteAddress](pAddress) <
        cast[ByteAddress](pExportDirectory) + pDataDirectory.Size:

        var hForward: HMODULE

        ##  pAddress is equal to DLL.FUNC or DLL.#ORDINAL
        var c = $(cast[cstring](pAddress))

        if c == "":
            return nil

        pAddress = nil
        
        var tmp = c.split('.')
        var dllName = tmp[0] & ".dll"
        var funcName = tmp[1]

        var num: int
        if funcName[0] == '#':
            num = parseInt(funcName[1..^1])
        else:
            num = CalcHashSalt(funcName)
        
        pLoadLibraryA = cast[LoadLibraryA](GetExportAddress(GetModuleBaseAddress(CalcHashSalt("KERNEL32.DLL")), CalcHashSalt("LoadLibraryA")))

        if pLoadLibraryA == nil:
            return nil

        hForward = pLoadLibraryA(dllName)

        if hForward == 0:
            return nil

        pAddress = GetExportAddress(hForward, num)

    return cast[FARPROC](pAddress)
